[c10d] simplify the dump timeout logic and unify the async call

shuqiangzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120331

Summary:
The current dump timeout logic is a bit cumbersome as it needs 2 times: 1.
timeout, 2. wake up time. And in theory the caller just needs to wait
for a max of timeout value for the dump and declare the dump to be
either successful or not. Also we unify the async call using std::async
instead of a customized async lauch function for each operation.
Test Plan:
Unit tests
Reviewers:

Subscribers:

Tasks:

Tags:

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

shuqiangzhang committed:
[c10d] simplify the dump timeout logic and unify the async call
Summary:
The current dump timeout logic is a bit cumbersome as it needs 2 times: 1.
timeout, 2. wake up time. And in theory the caller just needs to wait
for a max of timeout value for the dump and declare the dump to be
either successful or not. Also we unify the async call using std::async
instead of a customized async lauch function for each operation.
Test Plan:
Unit tests
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] simplify the dump timeout logic and unify the async call"


Summary:
The current dump timeout logic is a bit cumbersome as it needs 2 times: 1.
timeout, 2. wake up time. And in theory the caller just needs to wait
for a max of timeout value for the dump and declare the dump to be
either successful or not. Also we unify the async call using std::async
instead of a customized async lauch function for each operation.
Test Plan:
Unit tests
Reviewers:

Subscribers:

Tasks:

Tags:

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] simplify the dump timeout logic and unify the async call"


Summary:
The current dump timeout logic is a bit cumbersome as it needs 2 times: 1.
timeout, 2. wake up time. And in theory the caller just needs to wait
for a max of timeout value for the dump and declare the dump to be
either successful or not. Also we unify the async call using std::async
instead of a customized async lauch function for each operation.
Test Plan:
Unit tests
Reviewers:

Subscribers:

Tasks:

Tags:

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] simplify the dump timeout logic and unify the async call"


Summary:
The current dump timeout logic is a bit cumbersome as it needs 2 times: 1.
timeout, 2. wake up time. And in theory the caller just needs to wait
for a max of timeout value for the dump and declare the dump to be
either successful or not. Also we unify the async call using std::async
instead of a customized async lauch function for each operation.
Test Plan:
Unit tests
Reviewers:

Subscribers:

Tasks:

Tags:

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] simplify the dump timeout logic and unify the async call"


Summary:
The current dump timeout logic is a bit cumbersome as it needs 2 times: 1.
timeout, 2. wake up time. And in theory the caller just needs to wait
for a max of timeout value for the dump and declare the dump to be
either successful or not. Also we unify the async call using std::async
instead of a customized async lauch function for each operation.
Test Plan:
Unit tests
Reviewers:

Subscribers:

Tasks:

Tags:

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -945,69 +945,12 @@ void ProcessGroupNCCL::enableCollectivesTiming() {
-    // care about the retval
-    try {
-      fut.get();
-      std::this_thread::sleep_until(wakeUpTime);
```
Let me try to explain the reason for the old logic and then see if the new logic is adequate.

Timeout - we wait up this long for a dump before abandoning. Ideally we never abandon and the future completed well before the timeout expires 

Sleep- even in cases the local dump was fast, we wait a minimum time before letting our process exit such that mast won't kill other processes that may be still dumping.

Controversy- if we set the sleep too long it means job-restart time is impacted. We want jobs to restart asap to improve goodput.  But if we set sleep too short we may observe lost dump data from ranks killed by mast, and we want to never lose dump data.

shuqiangzhang replied:
Thanks for explaining the original logic here!
Then there seems to be some duplications of logic here. E.g., In monitor thread, there is another 'sleep' serving the same purpose of waiting for the rank/or other ranks to dump `// Leave another two mins for desync report generation or process group
    // destroy.
    std::this_thread::sleep_for(std::chrono::seconds(heartbeatTimeoutInSec_));` Feel its better to remove this sleep logic from the general function of 'waitForFutureOrTimeout'. And leave this 'sleep' logic only in the last thread to join (monitoring thread).

wconstab replied:
Ok that sounds good. I forgot if we still have a separate code path to dump from watchdog thread when collective timeout happens? Then it needs to sleep too.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1568,22 +1515,22 @@ void ProcessGroupNCCL::watchdogHandler() {
+              waitForFutureOrTimeout(
+                  *optAsyncDebugDump,
+                  std::chrono::milliseconds(waitTimeoutDumpInMilSec_),
+                  "Flight recorder dump in watchdog");
```
I think we're missing the extra 2 min wait in this path? 

Also I think it would be better if we only had one path. Perhaps in another PR, refactor to let watchdog simply signal timeout occurred, and monitor thread to then exercise the dumping logic?

shuqiangzhang replied:
Agree, checking the code path, watchdog thread can throw an exception immediately after detecting the timeout. will add a sleep here for now. Also agree that we probably should not dump from 2 different places. But kind of understand it why it is there now, watchdog thread to detect timeout on local rank, monitoring thread to detect timeout on remote ranks...

shuqiangzhang committed:
Update on "[c10d] simplify the dump timeout logic and unify the async call"


Summary:
The current dump timeout logic is a bit cumbersome as it needs 2 times: 1.
timeout, 2. wake up time. And in theory the caller just needs to wait
for a max of timeout value for the dump and declare the dump to be
either successful or not. Also we unify the async call using std::async
instead of a customized async lauch function for each operation.
Test Plan:
Unit tests
Reviewers:

Subscribers:

Tasks:

Tags:

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] simplify the dump timeout logic and unify the async call"


Summary:
The current dump timeout logic is a bit cumbersome as it needs 2 times: 1.
timeout, 2. wake up time. And in theory the caller just needs to wait
for a max of timeout value for the dump and declare the dump to be
either successful or not. Also we unify the async call using std::async
instead of a customized async lauch function for each operation.
Test Plan:
Unit tests
Reviewers:

Subscribers:

Tasks:

Tags:

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wconstab reviewed (approved):
Lgtm!

shuqiangzhang committed:
Update on "[c10d] simplify the dump timeout logic and unify the async call"


Summary:
The current dump timeout logic is a bit cumbersome as it needs 2 times: 1.
timeout, 2. wake up time. And in theory the caller just needs to wait
for a max of timeout value for the dump and declare the dump to be
either successful or not. Also we unify the async call using std::async
instead of a customized async lauch function for each operation.
Test Plan:
Unit tests
Reviewers:

Subscribers:

Tasks:

Tags:

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] simplify the dump timeout logic and unify the async call"


Summary:
The current dump timeout logic is a bit cumbersome as it needs 2 times: 1.
timeout, 2. wake up time. And in theory the caller just needs to wait
for a max of timeout value for the dump and declare the dump to be
either successful or not. Also we unify the async call using std::async
instead of a customized async lauch function for each operation.
Test Plan:
Unit tests
Reviewers:

Subscribers:

Tasks:

Tags:

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

shuqiangzhang commented:
Further fixing the already broken UT before this PR due to collective nature of commAbort:
python test/distributed/test_c10d_nccl.py -k test_timeout_dumps_timing_enabled_False

shuqiangzhang committed:
Update on "[c10d] simplify the dump timeout logic and unify the async call"


Summary:
The current dump timeout logic is a bit cumbersome as it needs 2 times: 1.
timeout, 2. wake up time. And in theory the caller just needs to wait
for a max of timeout value for the dump and declare the dump to be
either successful or not. Also we unify the async call using std::async
instead of a customized async lauch function for each operation.
Test Plan:
Unit tests
Reviewers:

Subscribers:

Tasks:

Tags:

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] simplify the dump timeout logic and unify the async call"


Summary:
The current dump timeout logic is a bit cumbersome as it needs 2 times: 1.
timeout, 2. wake up time. And in theory the caller just needs to wait
for a max of timeout value for the dump and declare the dump to be
either successful or not. Also we unify the async call using std::async
instead of a customized async lauch function for each operation.
Test Plan:
Unit tests
Reviewers:

Subscribers:

Tasks:

Tags:

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

shuqiangzhang commented:
@pytorchbot merge
